doc/KUKSA.val_data_broker/authorization.md

```diff
@@ -0,0 +1,505 @@
+resource owner itself, or what kind of access rights a certain group membership or role would provide. What this means in practice is that anyone authorized to get a valid access token is assumed to have full access rights within KUKSA.VAL, which will only be limited by the scope provided.
+
+There is nothing in the design preventing KUKSA.VAL from including support for other types
+of verifications in the future. The scope as specified here will continue to work as a way to limit the access granted by these other means anyway.
+
+
+### Limit access to resources
+The issued access token will typically be intended to for use with a certain resource. In the
+context of KUKSA.VAL, this can be a certain vehicle, or a certain backend service.
+
+To prevent access tokens issued for one resource from being used to gain access to other,
+unintended resources, OAuth2 introduces the concept of [Resource Indicators [RFC 8707]](https://datatracker.ietf.org/doc/html/rfc8707). By specifying the intended resource in the request
+to the authorization server, the resulting `"aud"` or audience claim should be populated with
+this resource.
+
+From [RFC9068](https://datatracker.ietf.org/doc/html/rfc9068#JWTATLRequest):
+> If the request includes a "resource" parameter (as defined in [[RFC8707](https://datatracker.ietf.org/doc/html/rfc6749)]), the resulting JWT access token "aud" claim SHOULD have the same value as the "resource" parameter in the request.
+
+and from [RFC8707](https://datatracker.ietf.org/doc/html/rfc6749):
+> To prevent cross-JWT confusion, authorization servers MUST use a distinct identifier as an "aud" claim value to uniquely identify access tokens issued by the same issuer for distinct resources. For more details on cross-JWT confusion, please refer to Section 2.8 of [RFC8725].
+
+By using a resource identifier that uniquely identifies the resource (i.e. a certain vehicle),
+and including this in the `"aud"` claim, KUKSA.VAL can verify that the access token isn't meant
+for another resource.
+
+KUKSA.VAL also needs to verify that the access token is intended to be used for KUKSA.VAL.
+An `"aud"` claim that would address both these requirements could look something like this (TBD):
+
+```yaml
+{
+  "aud": [
+    "vehicle/5GZCZ43D13S812715/kuksa.val.v1"
+  ],
+  ...
+}
+```
+
+where `5GZCZ43D13S812715` could be the VIN or another unique identifier available to KUKSA.VAL
+for local verification.
+
+### Limit access with scope
+The scope(s) used in in KUKSA.VAL to limit access rights, utilize "prefix scopes" that
+consist of an action prefix followed by an (optional) VSS path. This provides a flexible
+way to limit for which VSS signals the action is allowed.
+
+| Scope                 | Description                 |
+|-----------------------|-----------------------------|
+|`<ACTION>[:<PATH>]`    | Allow ACTION for PATH       |
+
+Examples of KUKSA.VAL scope strings:
+
```

Comment:
Ok, as promised, "scopes", I feel this is the last "big thing" we need to align on. Understanding, down the road we can always add new scopes, what is the initial set of useful scopes.

let me mainly refer to what is in the current draft, and also the reply here https://github.com/eclipse/kuksa.val/pull/471#discussion_r1096407754

I sense a consensus, that `read:<something>` is a good idea and well understood.

from the previous comment
> I interpret most of the proposed changes as a desire for a more direct mapping between scope and the current implementation of kuksa.val.v1 where setting a field is the API.

I can follow this line of argument, and actually my first feeling on reading you proposal was, you intend to align on some "eventually-to-be" API wis provide, while I was going the more generic route with saying what to read/write. That was mainly influenced by checking i.e. the Github scopes our other Web APIs where, quite often scopes revolve a lot around "read:that" or "write:this". 

But I agree, this is about intent. So let's check the other intents of our users/clients: 

Another thing an application likely wants to do is "actuate", that is trying to influence the state of something in a vehicle. I would agree, that `actuate:something` is a good scope for that. My current understanding would be, that this implies, in terms of current API/data model, that a client is able to write the target value, and also read it (I would just assume restricting "read" here, does not make much sense), and possibly also being able to read the current value.


That leaves the "southbound things". We are converging on calling these things providers, see discussion in #479 and notes in https://github.com/boschglobal/kuksa.val/wiki/Terminology . There is a difference beween

 - Providing sensor values: E.g. that is what DBC feeder does, or really is the definition of what have been called "feeders" before. I think the scope must make that clear, and not only say provide. I would suggest `provide-sensor` (in the current version of this draft there seems no difference, the wiki page suggests `provide:value`, but syntax-wise, I think it is more elegant if ":" always separates Path from action and not "sub-actions". `provide-sensor` would imply - in current API terms - to be able to write the current value of a sensors. I would assume that also applies being able to read it, although I thin it does not matter so mich
 - Providing services to actuate things. I suggest calling the scope `provide-actuator` , the wiki suggests, `provide:actuation`, argument for ":" applies as above. And I feel using the exact term "actuator" as sensor in VSS makes things consistent, and what in fact you do is providing an actuator to the SDV world. Currently this is probably identical to "read", but having a separate scope, can be useful later, in case there will be a more synchronous/special API  to report on actuation success/errors etc.

I am not convinced we should have an aggregate `provide` scope for both aspects (now, could still be introduced later). Currently an example for a component needed both is our seat service example: It moves the seat (`provide-actuation`) and reports back (`provide-sensor`), however I would assume (can't prove it though :) ), that in real deployments both aspects would often be handled by different components, just due to the different safety requirements. maybe the actuation services calls a SomeIP service, or some exiting lower level API, while positions are on CAN/Ethernet somewhere anyway and can be parsed directly and cheaply in QM domain anyway.  Also it would leave the road open to other `provide-something`capabilites, although I can't think of any right now.

last point, which I would like to at least define in initial scopeset: Modification of VSS tree. While we may not implement fine-grained modification of every VSS metadata aspect via API in the short term, I do think the overlay concept, i.e. the ability to split a large VSS model into different parts is an important/powerful aspect of the VSS ecosystem. We already support loading several models during startup, I could see us extending this to a runtime API soonish, where you might instruct db to load additional JSONs via an API call (assume, some new/specific <del>feeder</del>provider is deployed for an app, it might want to "register" datapoinnts/extend the model). I think it wold be relatively straight forward, to limit which parts of the VSS tree an app is allowed to touch. For ease of use, maybe start with the `modify` mentioned below (I fell that "includes" create). I think just the term is a bit to generic, maybe `modify-model` is better

tldr; my suggested "initial actions" for scopes are
 - `read`
 - `provide-sensor`
 - `provide-actor`
 - `modify-model`

Reply:
> I would suggest `provide-sensor`

I think this is a very problematic choice of wording.

> So provide-sensor might be better catching the intent and would imply - in current API terms - to be able to write the current value of a signal. I would assume that also applies being able to read it, although I think it does not matter so mich

On the contrary, I would say it _fails_ to catch the intent and is even misleading. Using `provide-sensor` fails to convey that this scope grants authorization to provide values for any type of VSS signal, be it an `Attribute`, `Actuator` or `Sensor`.

> I think -value may be accused of being to reliant on what the current API is. Just trying to use your arguments against you :)

The _value_ of a signal is established VSS terminology (right?), and not specific to the current API(s).

> in the current version of this draft there seems no difference, the wiki page suggests provide:value, but syntax-wise, I think it is more elegant if ":" always separates Path from action and not "sub-actions"

Did you see (or miss) this part of the draft?

>| Subactions for `read` | Description                        |
>|-----------------------|------------------------------------|
>| `read:field:<FIELD>`  | Allow client to only read a certain field for matching signals |
>
>| Subactions for `provide` | Description                     |
>|--------------------------|---------------------------------|
>| `provide:value` | Allow client to provide the value (but not respond to actuation requests) for matching signals. (includes `read`) |
>| `provide:actuation` | Allow client to respond to actuation requests (but not provide the value) for matching signals. (includes `read`) |
>
>| Subactions for `edit` | Description                     |
>|-----------------------|---------------------------------|
>| `edit:field:FIELD` | Allow client to edit metadata field FIELD for matching signals. (includes `read`) |
>
>| Subactions for `edit` | Description                     |
>|-----------------------|---------------------------------|
>| `edit:field:FIELD` | Allow client to edit metadata field FIELD for matching signals. (includes `read`) |

I will update the document with a better description of the thoughts behind structuring the scopes like this. In short, the `:` is used to describe a _hierarchical access right_ structure.

### Hierarchical Access Right
The core of the concept is that the top of the access rights hierarchy includes all the access rights below it.

So for example, `read` is the top (or root) in an access rights hierarchy including read access to everything. `read:Vehicle.Speed` is a child in that hierarchy and is included in both `read:Vehicle` and `read`. Furthermore, this can be expanded to include `read:field:FIELD` or `read:FIELD` which can be combined with either paths or tags (also separated by `:`) to further specify the specific child access right in this hierarchy.

Analogous to this, `provide` is the root in another access rights hierarchy. Both `provide:value` and `provide:actuation` is included in this root.

Your argument:

> I would assume (can't prove it though :) ), that in real deployments both aspects would often be handled by different components

This doesn't make an access rights hierarchy invalid or harder to use for the specified use case. Even though these aspects would _often_ be handled by separate components, using an hierarchy like this will make it easy to describe the access rights needed for components that provide **both** values and actuation.

Reply:
> I suggest calling the scope provide-actuator , the wiki suggests, provide:actuation, argument for ":" applies as above. And I feel using the exact term "actuator" as sensor in VSS makes things consistent, and what in fact you do is providing an actuator to the SDV world.

Again, I think this terminology is very misleading.

Using the term _actuator_ to only refer to the actuation part is **not** consistent with VSS terminology.
Using the term _sensor_ to refer to the value of a signal is **not** consistent with VSS terminology.

An `Actuator` in VSS terminology includes _both_ the ability to set and read it. A provider providing only one of these features is not providing a complete actuator "to the SDV world".

Reply:
> The value of a signal is established VSS terminology (right?), and not specific to any current API.
I think that my be the point where I do not fully agree. 

We think about it in this terms in KUKSA.val (and VISS partially shares the same worldview). Sensors can be read, actuators can be, well, actuated. 
VSS itself however is just a data model, and does not say anything about applications. A signal has a name and a type. If you e.g. use this model in your data warehouse/backend, then something being an "actuator" might have some relevance there (maybe counting how often it has been used), but a (current) value might be not at all relevant here. Similar for sensors, maybe only statistical moments are stored. If VSS is used in an authorisation server, maybe "actuator" as VSS type is only a flag making sure, certain apps never get to use it.

(I know there are hints that might lead you otherwise e.g. _examples_ how you might want to use it here https://covesa.github.io/vehicle_signal_specification/rule_set/data_entry/sensor_actuator/ or this things even called Value restrictions "https://covesa.github.io/vehicle_signal_specification/rule_set/data_entry/allowed/").

## Suggestions
Therefore, even taking all explanatory text around VSS as normative, I feel `provide:value` to generic and I would have no good intuition what it means.

Would it be possible to make clear it `provide:currentvalue` or `provide-currentvalue` (wrt to ":" not yet sure if provide hierarchy has that much practical use, but I can agree it,  does not hurt)
 or alternatively in line with current  terminology revision `provide:state`
I would interpret all of these to be able to "write" the current value for sensors/actuators/attributes 

Also I think, in my shortlist I forgot  `actuate` for VSS consumers/users waning to "write" an target_value. It hink that is fine as is in the document.

That leaves the scope for providers providing actuation in a vehicle. As I just called this thing `actuation-provider` in terminology I would yield, and accept `provide:actuation` as the better solution. My understanding would be this will allow reading (at least) target_value, but likely everything as "read", but not allow writing "currentvalue".

So my new shortlist would be

```
read
actuate
provide:state or provide:currentvalue
provide:actuation
modify-model
```

Reply:
> Therefore, even taking all explanatory text around VSS as normative, I feel provide:value to generic and I would have no good intuition what it means.

On that I completely agree.

In the terminology section, I am hesitant to use `Value provider` as it is, just like you say, to generic and the meaning not immediately obvious (it sounds more like something out of a PR slide). `Data provider` is proposed as an alternative there, so I guess `provide:data` could be another item to add to the list of alternatives to ponder. I even think that term was used spontaneously at FOSDEM (by some external party).

Reply:
Regarding `modify-model`.

I haven't fully articulated why it could make sense to split this functionality into separate access rights (or actions). This is what it looks like in the draft at the moment.

| Action    | Description                                   |
|-----------|-----------------------------------------------|
| `edit`    | Allow client to edit metadata of a VSS entry. |
| `create`  | Allow client to create a VSS entry (under a certain path). If a VSS entry already exists, `edit` is needed to change it. |

The reason for differentiating between creating and editing entries is that creating an entry shouldn't have a huge effect on already running applications in the vehicle. Editing an entry on the other hand, could have a huge effect as it has the potential to invalidate the state of a lot of applications. That's why the right to create an entry doesn't automatically include the authorization to edit it.

I think this is an area where we need to think a bit to get a sensible implementation.

Regarding naming:
`create` seems intuitive enough, e.g. `create:Vehicle.Speed` or `create:Private.*`.

`edit` might be a bit to generic

Reply:
I fully agree with create as a way do allow "adding" things to the model. My assumption is that the create scope will not allow "overwriting" or changing existing parts in the model. That is useful if you are composing your model during runtime, to ensure "Nothing goes wrong"

I guess "modify-model" was my way to say "edit". One common approach VSS side could be, that you load a base model, for example the VSS 3.0 and on top of that load an overlay, that will set the "defaults" for all attributes and attributes like min/may value for a specific vehicle. That something like edit rights, and I was not sure if that _only_ allows editing existing leafs and branches or _also_ extending them.

I was searching for a scope that basically says "Vehicle.ADAS" is yours, do whatever. And as the most generic I was thinking it might be good to have implemented first.

But maybe the answer is, we do need to start with both
`create:` to _add_ but not change, `modify:` (or edit) to change/ammend existing nodes. 

Or build another hierarchy? 
 - `modify:create`  
 - `modify:amend`
 -  `modify:change`

Reply:
I think `modify` and `create` are so different that it would cause more issues than benefits to try to combine them. The combination `modify:change` is perhaps a good example of this :)

Regarding `modify` / `edit` / `update` or whatever we want to call it, a few questions needs to be answered (I think) before we decide to add support for it:
* Would this scope allow changing the `entry type` of an already existing entry.
* Would this scope allow changing the `data type` of an already existing entry.

Maybe such changes should be disallowed completely. What other changes would be problematic?

These issues don't exist with regard to `create`, and my gut feeling is that a good way forward could be to start by adding `create`, if that is something which is needed.

Reply:
Yes, I think on create I think we have a consensus, that it is useful and what it means, so

```
read
actuate
provide:data
provide:actuation
create
```

wrt to "edit/modify" yes I think changing type/datatype/unit are "dangerous" operations (not implying we should not allow it, but maybe needs different scopes).

One near-term usecase I see for something where `create` is not enough is something which in VSS community is sometimes referred to "deployment information", most simple example: Your VSS model has attributes, most of them are by nature quite static (e.g. VIN), so you provide an overlay that modifies "default", i.e. the value for them (and defaults have been removed from VSS see https://github.com/COVESA/vehicle_signal_specification/pull/284)
The same goes for deployment specific information such as min/max

You can do this of course "pre-deployment", i.e. providing an adapted VSS file already, but there might be reasons somebody wants to do it "online", because that basically follows the pattern"config bits" you already have today in some ECUs enable/disable features or do certain settings.

I would have more (far-fetched) examples. Doing instances in vehicle would be an even more invasive operation.

My wish would be, that we do have some scope that might enable at least the first usecase. Either declaring it a "dangerous" operation and note you should use with care, or maybe limit changing "dangerous" things listed above. Any idea how to capture this in a obvious, self-explanatory, useful scope?

Reply:
> Any idea how to capture this in a obvious, self-explanatory, useful scope?

Not spontaneously, no. My initial thought was:

What you are describing is just an indirect way to change the value of an attribute. That functionality is already covered by the `provide` scope, i.e. `provide:Vehicle.TheAttribute`.

> and defaults have been removed from VSS see https://github.com/COVESA/vehicle_signal_specification/pull/284

It's only removed for sensors & actuators, right? It's still there for attributes, as was the original intention, right.

Changing the default value wouldn't even be persisted in the current implementation. Maybe that's another dimension to think about (runtime changes vs persistant changes). And maybe the dangerous things wouldn't be so dangerous if they only applied to an offline "configuration" and where not applied immediately (to a currently running instance). I don't know, thinking out loud here.